File: flutter_app/test/portfolio_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:pocketvault/domain/models/enums.dart';
import 'package:pocketvault/domain/models/owned_card.dart';
import 'package:pocketvault/domain/services/portfolio.dart';
import 'package:pocketvault/domain/services/pricing.dart';

/// Tests de la logique métier pure : aucune dépendance Flutter/UI, ce qui
/// valide l'isolation de la couche domaine.
MarketQuote quote(int spot) => MarketQuote(
      cardId: 'x',
      spotCents: spot,
      byCondition: {
        for (final c in Condition.values) c: (spot * c.multiplier).round(),
      },
      byGrade: const {},
    );

OwnedCard owned({
  Condition condition = Condition.nearMint,
  Grade grade = const Grade(company: GradingCompany.raw),
  int acquired = 0,
  String uid = 'u',
  String cardId = 'x',
}) =>
    OwnedCard(
      uid: uid,
      cardId: cardId,
      condition: condition,
      grade: grade,
      acquiredCents: acquired,
      acquiredAt: DateTime(2024),
      binderIds: const [],
      favorite: false,
    );

void main() {
  group('Pricing', () {
    test('décote d\'état pour carte brute', () {
      expect(
        Pricing.valueOf(owned(condition: Condition.lightlyPlayed), quote(10000)),
        7800,
      );
    });

    test('multiplicateur PSA 10', () {
      expect(
        Pricing.valueOf(
          owned(grade: const Grade(company: GradingCompany.psa, score: 10)),
          quote(10000),
        ),
        75000,
      );
    });

    test('plus-value latente', () {
      final pnl = Pricing.unrealizedPnl(
        owned(
          grade: const Grade(company: GradingCompany.psa, score: 10),
          acquired: 20000,
        ),
        quote(10000),
      );
      expect(pnl.cents, 55000);
      expect(pnl.ratio, closeTo(2.75, 1e-9));
    });
  });

  group('Portfolio', () {
    test('agrège valeur, coût et top carte', () {
      final cards = [
        owned(uid: 'a', acquired: 1000),
        owned(
          uid: 'b',
          cardId: 'y',
          grade: const Grade(company: GradingCompany.psa, score: 9),
          acquired: 5000,
        ),
      ];
      MarketQuote? lookup(String id) => quote(id == 'y' ? 8000 : 2000);

      final stats = Portfolio.compute(cards, lookup);
      expect(stats.totalCostCents, 6000);
      expect(stats.totalValueCents, 2000 + (8000 * 3.1).round());
      expect(stats.topCardUid, 'b');
    });

    test('seriesDelta mesure la variation', () {
      final d = Portfolio.seriesDelta([100, 150]);
      expect(d.cents, 50);
      expect(d.ratio, closeTo(0.5, 1e-9));
    });
  });
}

File: src/domain/services/__tests__/portfolio.test.ts
import { computePortfolioStats, seriesDelta } from '../portfolio';
import { unrealizedPnl, valueOfOwnedCard } from '../pricing';
import { MarketQuote, OwnedCard } from '../../types';

/**
 * Tests de la logique métier pure. Aucune dépendance React/native : la
 * couche domaine est entièrement vérifiable en isolation — c'est tout
 * l'intérêt de la séparation logique/UI.
 */

function quote(spotCents: number): MarketQuote {
  return {
    cardId: 'x',
    currency: 'EUR',
    spotCents,
    byCondition: {
      NearMint: spotCents,
      LightlyPlayed: Math.round(spotCents * 0.78),
      Played: Math.round(spotCents * 0.55),
      Poor: Math.round(spotCents * 0.3),
    },
    byGrade: {},
    updatedAt: 0,
  };
}

function owned(partial: Partial<OwnedCard>): OwnedCard {
  return {
    uid: 'u',
    cardId: 'x',
    condition: 'NearMint',
    grade: { company: 'Raw', score: null },
    acquiredCents: 0,
    acquiredAt: 0,
    binderIds: [],
    favorite: false,
    ...partial,
  };
}

describe('pricing', () => {
  it('applique la décote d état pour une carte brute', () => {
    const card = owned({ condition: 'LightlyPlayed' });
    expect(valueOfOwnedCard(card, quote(10000))).toBe(7800);
  });

  it('applique le multiplicateur de gradation PSA 10', () => {
    const card = owned({ grade: { company: 'PSA', score: 10 } });
    expect(valueOfOwnedCard(card, quote(10000))).toBe(75000);
  });

  it('calcule une plus-value latente positive', () => {
    const card = owned({
      grade: { company: 'PSA', score: 10 },
      acquiredCents: 20000,
    });
    const pnl = unrealizedPnl(card, quote(10000));
    expect(pnl.cents).toBe(55000);
    expect(pnl.ratio).toBeCloseTo(2.75);
  });
});

describe('portfolio', () => {
  it('agrège valeur, coût et carte la plus chère', () => {
    const cards = [
      owned({ uid: 'a', acquiredCents: 1000, condition: 'NearMint' }),
      owned({
        uid: 'b',
        cardId: 'y',
        grade: { company: 'PSA', score: 9 },
        acquiredCents: 5000,
      }),
    ];
    const lookup = (id: string) => quote(id === 'y' ? 8000 : 2000);

    const stats = computePortfolioStats(cards, lookup);
    expect(stats.totalCostCents).toBe(6000);
    // a: 2000 (NM) ; b: 8000 * 3.1 = 24800
    expect(stats.totalValueCents).toBe(26800);
    expect(stats.topCardUid).toBe('b');
    expect(stats.unrealizedCents).toBe(20800);
  });
});

describe('seriesDelta', () => {
  it('mesure la variation entre premier et dernier point', () => {
    const d = seriesDelta([{ cents: 100 }, { cents: 150 }] as never);
    expect(d.cents).toBe(50);
    expect(d.ratio).toBeCloseTo(0.5);
  });
});
